armory/scenarios/scenario.py

```diff
@@ -0,0 +1,423 @@
+            if config["model"]["fit"]:
+                config["model"]["fit_kwargs"]["nb_epochs"] = 1
+            if config.get("attack", {}).get("type") == "preloaded":
+                config["attack"]["check_run"] = True
+            # For poisoning scenario
+            if config.get("adhoc") and config.get("adhoc").get("train_epochs"):
+                config["adhoc"]["train_epochs"] = 1
+        self._check_config_and_cli_args(
+            config, num_eval_batches, skip_benign, skip_attack, skip_misclassified
+        )
+        self.config = config
+        self._set_output_dir(self.config)
+        self.num_eval_batches = num_eval_batches
+        self.skip_benign = bool(skip_benign)
+        self.skip_attack = bool(skip_attack)
+        self.skip_misclassified = bool(skip_misclassified)
+        if skip_benign:
+            logger.info("Skipping benign classification...")
+        if skip_attack:
+            logger.info("Skipping attack generation...")
+        self.mongo_host = mongo_host
+        if self.mongo_host is not None:  # fail fast if pymongo is not installed
+            from armory.scenarios import mongo  # noqa: F401
+
+    def _set_output_dir(self, config: Config) -> None:
+        runtime_paths = paths.runtime_paths()
+        self.scenario_output_dir = os.path.join(
+            runtime_paths.output_dir, config["eval_id"]
+        )
+
+    def _check_config_and_cli_args(
+        self, config, num_eval_batches, skip_benign, skip_attack, skip_misclassified
+    ):
+        if skip_misclassified:
+            if skip_attack or skip_benign:
+                raise ValueError(
+                    "Cannot pass skip_misclassified if skip_benign or skip_attack is also passed"
+                )
+            elif "categorical_accuracy" not in config["metric"].get("task"):
+                raise ValueError(
+                    "Cannot pass skip_misclassified if 'categorical_accuracy' metric isn't enabled"
+                )
+            elif config["dataset"].get("batch_size") != 1:
+                raise ValueError(
+                    "To enable skip_misclassified, 'batch_size' must be set to 1"
+                )
+
+    def _load_estimator(self):
+        model_config = self.config["model"]
+        estimator, _ = config_loading.load_model(model_config)
+        self.predict_kwargs = model_config.get("predict_kwargs", {})
+        # TODO: handle fit_preprocessing_fn ???
+        return estimator
+
+    def _load_defense(self, estimator, train_split_default="train"):
+        model_config = self.config["model"]
+        defense_config = self.config.get("defense") or {}
+        defense_type = defense_config.get("type")
+
+        if defense_type in ["Preprocessor", "Postprocessor"]:
+            logger.info(f"Applying internal {defense_type} defense to estimator")
+            estimator = config_loading.load_defense_internal(defense_config, estimator)
+
+        if model_config["fit"]:
```

Comment:
for the sake of modularity, would it make sense to have a separate method encapsulating model training? `_load_defense()` isn't the first place I'd expect training to live

Reply:
Yes, I agree.